### ISSUE TITLE
Linking to statically compiled z3 needs additional libraries on Linux

### ISSUE THREAD

**Issue by quicquid** [Z3Prover/z3#2211]
I have posted this question first on [Stack Overflow](https://stackoverflow.com/questions/55402008/linking-to-statically-compiled-z3-needs-additional-libraries-on-linux) but perhaps someone here might be interested as well.

I am using cmake to build Z3 statically (usually directly from master, but I'm sometimes a few commits behind) on Linux. This works well, but during linking I get the following message:

    /usr/bin/ld: /usr/lib/gcc/x86_64-redhat-linux/8/libgomp.a(target.o): in function `gomp_target_init':
    (.text+0x32c): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking

Accordingly, I have to add `-lgomp -pthread -lrt -ldl` whenever I want to link against z3. Is there a way I can check if the libraries are part of the library? A user on SO also suggested that mixed linking is neccessary for some setups -- is there a way I can make sure no dynamic libraries are picked up during linking?

I will put the exact configuration options and a test program into a comment.

**Comment by quicquid** [comment]
The environment is Fedora Linux with static libraries located at:

```
for I in gomp rt pthread dl; do locate lib$I.a;done
/usr/lib/gcc/x86_64-redhat-linux/8/libgomp.a
/usr/lib/gcc/x86_64-redhat-linux/8/32/libgomp.a
/usr/lib64/librt.a
/usr/lib64/libpthread.a
/usr/lib64/libdl.a
```

I set up cmake for z3 with the ollowing options:

```cmake -DBUILD_LIBZ3_SHARED=false -DCMAKE_INSTALL_PREFIX=/opt/z3-devel -G "Unix Makefiles" ../```

The compilation command is:

    g++ -c -I /opt/z3-devel/include -static -o issue.o issue.cc

Linking is done with

    g++ -static -L /opt/z3-devel/lib64 -o issue issue.o -lz3

and 

    g++ -static -L /opt/z3-devel/lib64 -o issue issue.o -lz3 -lgomp -pthread -lrt -ldl

respectively. The file itself is just a simple application running the Fourier Motzkin tactic (file attached).

**Comment by wintersteiger** [comment]
This is definitely not just your problem, but most users are happy with dynamic linkage. In addition to my comment on Stack Overflow:  the warning message reads `warning: Using 'dlopen' ...`, i.e. your copy of `libgomp.a` contains a reference to the function that opens dynamic libraries, which will require dynamic system libraries. You can try to compile a completely static version of `libgomp.a`, but I think it may not be worth the effort. I haven't looked around though, maybe someone has already done that elsewhere.

**Comment by quicquid** [comment]
Thank you, that's good to know -- the background for my question is that I am currently creating a cmake based build system for software that depends on z3. So far, I just had the additional flags in the custom Makefile but I'd like to make sure this works in other environments as well. I'll see if I can get my hands on an Ubuntu installation and see if the same libraries are necessary.

**Comment by delcypher** [comment]
@quicquid If you're building a project that depends on Z3 and it uses CMake as the build system you should take a look at https://github.com/Z3Prover/z3/blob/master/examples/c%2B%2B/CMakeLists.txt . We have special support there that imports the Z3 library as an "imported target". It should any of the necessary link dependencies of a static libz3 archive.

**Comment by quicquid** [comment]
@delcypher Thank you, I will have a look! Do you know if the `find_package(Z3,...)`statement refers to the one contained in [Seahorn](https://github.com/seahorn/seahorn/blob/master/cmake/FindZ3.cmake)?

**Comment by delcypher** [comment]
@quicquid No it does not. There is an automatically generated `Z3Config.cmake` file that will appear in the root of the Z3 build directory when Z3 is built with CMake. The `Z3Config.cmake` is also installed if the install comes from a build and install of a CMake build of Z3. Also take a look at the CMake documention for `find_package()` because the example uses `CONFIG` mode which is different from CMake's module mode.

**Comment by NikolajBjorner** [comment]
omp has been deprecated
